### PR TITLE
Libc pthread Modify return value as per POSIX standard

### PR DESCRIPTION
**Pull request by thapav**
Modified return value and test cases for:
pthread_mutexattr_getprotocol()
pthread_mutexattr_setprotocol()
pthread_setcancelstate()
pthread_setcanceltype()
pthread_testcancel()

### CODE REVIEW

File: lib/libc/pthread/pthread_mutexattr_setprotocol.c [hunk 1]
```diff
@@ -80,25 +80,26 @@
  *
  ****************************************************************************/
 
-int pthread_mutexattr_setprotocol(FAR pthread_mutexattr_t *attr,
-				  int protocol)
+int pthread_mutexattr_setprotocol(FAR pthread_mutexattr_t *attr, int protocol)
 {
 	svdbg("attr=0x%p protocol=%d\n", attr, protocol);
-	DEBUGASSERT(attr != NULL);
```
Comment (jeongchanKim) [inline lib/libc/pthread/pthread_mutexattr_setprotocol.c]: why did you remove this line? even check with if statement, it is needed for debugging option

File: apps/examples/testcase/le_tc/kernel/tc_libc_pthread.c [hunk 2]
```diff
@@ -766,11 +766,14 @@ static void tc_libc_pthread_pthread_mutexattr_getprotocol(void)
 	TC_ASSERT_EQ("pthread_mutexattr_init", ret_chk, OK);
 
 #ifdef CONFIG_PRIORITY_INHERITANCE
+	attr.proto = PTHREAD_PRIO_INHERIT;
 	ret_chk = pthread_mutexattr_getprotocol(&attr, &mutexattr_protocol);
-	TC_ASSERT_EQ("pthread_mutexattr_getprotocol", ret_chk, PTHREAD_PRIO_INHERIT);
+	TC_ASSERT_EQ("pthread_mutexattr_getprotocol", ret_chk, OK);
```
Comment (jeongchanKim) [inline apps/examples/testcase/le_tc/kernel/tc_libc_pthread.c]: pthread_mutexattr_getprotocol doesn't return OK.
it returns attr->proto or PTHREAD_PRIO_NONE.

Comment (thapav) [inline apps/examples/testcase/le_tc/kernel/tc_libc_pthread.c]: These shall get and set the protocol attribute of a mutex attributes object pointed to by attr which was previously created by the function pthread_mutexattr_init().
Upon successful completion, the pthread_mutexattr_getprotocol() and pthread_mutexattr_setprotocol() functions shall return zero; otherwise, an error number shall be returned to indicate the error.
Reference LLink:
http://pubs.opengroup.org/onlinepubs/009695399/functions/pthread_mutexattr_getprotocol.html

File: lib/libc/pthread/pthread_mutexattr_setprotocol.c (hunk 1 above)
Comment (thapav) [inline lib/libc/pthread/pthread_mutexattr_setprotocol.c]: If it is required , I can remove my change.
As I am doing the task of code coverage where we require (code coverage>80%) , I replaced this statement with an if() statement to test one test case where attr=NULL.

Comment (thapav) [inline lib/libc/pthread/pthread_mutexattr_setprotocol.c]: Similar implementation is there for other pthread API's where we use svdbg instead of DEBUGASSERT(attr != NULL);

Comment (jeongchanKim) [inline lib/libc/pthread/pthread_mutexattr_setprotocol.c]: DEBUGASSERT is only working when CONFIG_DEBUG is enabled.
Most of the case, we disable the debug option, so I think it is needed

File: apps/examples/testcase/le_tc/kernel/tc_libc_pthread.c (hunk 2 above)
Comment (jeongchanKim) [inline apps/examples/testcase/le_tc/kernel/tc_libc_pthread.c]: yes.. but in our code, it returns attr->proto or PTHREAD_PRIO_NONE.
so we should check again. I will discuss with my team also which should I modify.

Comment (thapav) [inline apps/examples/testcase/le_tc/kernel/tc_libc_pthread.c]: ok

File: lib/libc/pthread/pthread_mutexattr_setprotocol.c (hunk 1 above)
Comment (thapav) [inline lib/libc/pthread/pthread_mutexattr_setprotocol.c]: When we run the test cases for pthread in TC mode, we cannot go forward with DEBUGASSERT as we get this error:
up_assert: Assertion failed at file:pthread/pthread_mutexattr_setprotocol.c line: 86 task: kerneltc
and the rest of the tc's are halted.
If it is not required under code coverage, do we leave out this test case:
ret_chk = pthread_mutexattr_setprotocol(NULL, protocol);
What do you suggest ?

Comment (jeongchanKim) [inline lib/libc/pthread/pthread_mutexattr_setprotocol.c]: If there is ASSERT for checking something NULL, you can ignore for checking that.

Comment (thapav) [inline lib/libc/pthread/pthread_mutexattr_setprotocol.c]: ok

**Comment by thapav** [comment]
Please review this one:
https://github.com/Samsung/TizenRT/pull/312/commits/c712d26b4d873e535b9582f3e404e5d723db74c6

File: apps/examples/testcase/le_tc/kernel/tc_libc_pthread.c (hunk 2 above)
Comment (jeongchanKim) [inline apps/examples/testcase/le_tc/kernel/tc_libc_pthread.c]: I think that we should modify the implementation of pthread_mutexattr_getprotocol API.
It is good for merging this TC after modifying

File: lib/libc/pthread/pthread_mutexattr_setprotocol.c [hunk 3]
```diff
@@ -80,25 +80,23 @@
...
 	svdbg("attr=0x%p protocol=%d\n", attr, protocol);
 	DEBUGASSERT(attr != NULL);
 
 #ifdef CONFIG_PRIORITY_INHERITANCE
-	if (protocol >= PTHREAD_PRIO_NONE &&
-				protocol <= PTHREAD_PRIO_PROTECT) {
-		attr->proto = protocol;
-		return OK;
-	}
+		if (protocol >= PTHREAD_PRIO_NONE && protocol <= PTHREAD_PRIO_PROTECT) {
```
Comment (sunghan-chang) [inline lib/libc/pthread/pthread_mutexattr_setprotocol.c]: remove one tab. From here, it has different (wrong) depth.

**Comment by thapav** [comment]
modified as per review comments

File: apps/examples/testcase/le_tc/kernel/tc_libc_pthread.c [hunk 4]
```diff
@@ -766,11 +766,14 @@ static void tc_libc_pthread_pthread_mutexattr_getprotocol(void)
...
+	TC_ASSERT_EQ("pthread_mutexattr_getprotocol", mutexattr_protocol, attr.proto);
 #else
 	ret_chk = pthread_mutexattr_getprotocol(&attr, &mutexattr_protocol);
-	TC_ASSERT_EQ("pthread_mutexattr_getprotocol", ret_chk, PTHREAD_PRIO_NONE);
+	TC_ASSERT_EQ("pthread_mutexattr_getprotocol", ret_chk, OK);
+	TC_ASSERT_EQ("pthread_mutexattr_getprotocol", mutexattr_protocol, attr.proto);
```
Comment (sunghan-chang) [inline apps/examples/testcase/le_tc/kernel/tc_libc_pthread.c]: When CONFIG_PRIORITY_INHERITANCE is disabled, attr's proto is not set. You can not compare with mutexattr_protocol and attr.proto.

File: apps/examples/testcase/le_tc/kernel/tc_libc_pthread.c [hunk 5]
```diff
@@ -1227,6 +1230,85 @@ static void tc_libc_pthread_pthread_rwlock_trywrlock(void)
...
+* @Scenario             :The function shall create a cancellation point in the calling thread
+                         It has no effect if cancelability is disabled.
+* @API'scovered         :pthread_testcancel
+* @Preconditions        :none
+* @Postconditions       :none
+* @return               :void
+*/
+static void tc_libc_pthread_pthread_testcancel(void)
+{
+
```
Comment (sunghan-chang) [inline apps/examples/testcase/le_tc/kernel/tc_libc_pthread.c]: remove unncessary empty line

File: lib/libc/pthread/pthread_mutexattr_setprotocol.c [hunk 6]
```diff
@@ -80,25 +80,23 @@
  *
  ****************************************************************************/
 
-int pthread_mutexattr_setprotocol(FAR pthread_mutexattr_t *attr,
-				  int protocol)
+int pthread_mutexattr_setprotocol(FAR pthread_mutexattr_t *attr, int protocol)
```
Comment (sunghan-chang) [inline lib/libc/pthread/pthread_mutexattr_setprotocol.c]: This file has changes for formatting and these changes are not matched with commit message. Let's split commits.

File: apps/examples/testcase/le_tc/kernel/tc_libc_pthread.c [hunk 7]
```diff
@@ -1227,6 +1230,85 @@ static void tc_libc_pthread_pthread_rwlock_trywrlock(void)
...
+* @Scenario             :The function shall atomically both set the calling thread's cancelability state to the indicated state 
+*                        and return the previous cancelability state at the location referenced by oldstate
+*                        If successful pthread_setcancelstate() function shall return zero;
+*                        otherwise, an error number shall be returned to indicate the error.
+* @API'scovered         :pthread_setcancelstate
+* @Preconditions        :none
+* @Postconditions       :none
+* @return               :void
+*/
+static void tc_libc_pthread_pthread_setcancelstate(void)
```
Comment (sunghan-chang) [inline apps/examples/testcase/le_tc/kernel/tc_libc_pthread.c]: There new tcs for cancel are not under conditional. Are they compiled normally?

Comment (sunghan-chang) [inline apps/examples/testcase/le_tc/kernel/tc_libc_pthread.c]: Is it working at both cases, enabled CONFIG_CANCELLATION_POINTS and disabled CONFIG_CANCELLATION_POINTS?

File: lib/libc/pthread/pthread_mutexattr_setprotocol.c (hunk 6 above)
Comment (thapav) [inline lib/libc/pthread/pthread_mutexattr_setprotocol.c]: okay

File: apps/examples/testcase/le_tc/kernel/tc_libc_pthread.c (hunk 5 above)
Comment (thapav) [inline apps/examples/testcase/le_tc/kernel/tc_libc_pthread.c]: done

File: apps/examples/testcase/le_tc/kernel/tc_libc_pthread.c (hunk 7 above)
Comment (thapav) [inline apps/examples/testcase/le_tc/kernel/tc_libc_pthread.c]: no it should work for disabled CONFIG_CANCELLATION_POINTS..i will modify

Comment (thapav) [inline apps/examples/testcase/le_tc/kernel/tc_libc_pthread.c]: tc_libc_pthread_pthread_setcanceltype & tc_libc_pthread_pthread_testcancel work under disabled CONFIG_CANCELLATION_POINTS
whereas, tc_libc_pthread_pthread_setcancelstate works normally

Comment (sunghan-chang) [inline apps/examples/testcase/le_tc/kernel/tc_libc_pthread.c]: No, pthread_setcanceltype and pthread_setcancelstate works both cases. But pthread_testcancel works only under enabled CONFIG_CANCELLATION_POINTS. (nop under disabled)

Comment (sunghan-chang) [inline apps/examples/testcase/le_tc/kernel/tc_libc_pthread.c]: Please give an answer for this.

Comment (thapav) [inline apps/examples/testcase/le_tc/kernel/tc_libc_pthread.c]: As discussed on chat, 
When CONFIG_CANCELLATION_POINTS is enabled 
os/kernel/task/ : task_setcanceltype.c & task_testcancel.c will be compiled
When CONFIG_CANCELLATION_POINTS is disabled 
lib/libc/sched/ : task_setcanceltype.c & task_testcancel.c will be compiled

Now, pthread_setcanceltype, pthread_setcancelstate , pthread_testcancel  use above mentioned task API's 

Hence, either we don't include any flag at all 
OR
We add 
under tc_libc_pthread.c >>
#ifndef CONFIG_CANCELLATION_POINTS 
        tc_libc_pthread_pthread_setcanceltype();
        tc_libc_pthread_pthread_testcancel();
#endif

under tc_pthread.c >>
#ifdef CONFIG_CANCELLATION_POINTS 
        tc_libc_pthread_pthread_setcanceltype();
        tc_libc_pthread_pthread_testcancel();
#endif

Comment (sunghan-chang) [inline apps/examples/testcase/le_tc/kernel/tc_libc_pthread.c]: I think second is better. TC was already split by location. If you do first, it makes location of TC wrong.

**Comment by sunghan-chang** [comment]
@thapav Could you share your plan for our comments?

File: apps/examples/testcase/le_tc/kernel/tc_libc_pthread.c (hunk 7 above)
Comment (thapav) [inline apps/examples/testcase/le_tc/kernel/tc_libc_pthread.c]: Okay i will make changes and update

Comment (thapav) [inline apps/examples/testcase/le_tc/kernel/tc_libc_pthread.c]: done
